metadata_templates.py

```diff
@@ -88,9 +173,65 @@ def metadata_templates_collection_gather(rule_args, callback, rei):
                 schemas.append(j)
             except Exception as e:
                 callback.writeLine('serverLog', '{}'.format(type(e)))
+        elif thetype == 'irods':
+            # get length of iRODS full logical path
+            ret = callback.msiObjStat(schema, irods_types.RodsObjStat())
+            objstat = ret['arguments'][1]
+            # get schema contents from iRODS full logical path
+            ret = callback.msiDataObjOpen(schema, 0)
+            fd = ret['arguments'][1] # iRODS file descriptor
+            ret = callback.msiDataObjRead(fd, objstat.objSize, irods_types.BytesBuf())
+            buf = ret['arguments'][2] # bytesbuf
+            byteslist = buf.get_bytes() # schema contents as list of bytes as integers
+            callback.msiDataObjClose(fd, 0)
+            # convert to json object
+            j = json.loads(bytes(byteslist).decode("utf-8"))
+            # add to schemas array
+            schemas.append(j)
+        elif thetype == 'local':
+            # get contents
+            with open(schema, 'r') as f:
+                content = f.read()
+                # convert to json object
+                j = json.loads(content)
+                # add to schemas array
+                schemas.append(j)
         else:
             callback.writeLine('serverLog', 'Type [{}] Not Supported By Metadata Templates'.format(thetype))
+
+    # if recursive (and not the root),
+    # get all schema locations attached to this collection's parent collection, recursively
+#    callback.writeLine('serverLog', 'metadata_templates_collection_gather: before recursive flag')
+#    callback.writeLine('serverLog', 'length of schemas [{0}]'.format(len(schemas)))
+    if int(recursive):
+        callback.writeLine('serverLog', 'metadata_templates_collection_gather: inside recursive flag [{0}]'.format(logical_path))
+
+        clean_logical_path = logical_path.rstrip('/')
```

Comment:
The result of this operation is thrown away if any of the conditions on lines 210-214 are true.

You could change L216 to be an else-block and put this and the original logic starting on L216 inside the else-block. For example:
```py
else: # This is the new L216.
    clean_logical_path = logical_path.rstrip('/')
    if clean_logical_path.count('/') == 1:
        # ...
    else:
        # ...
```

Reply:
yes, have refactored to clean it once at the beginning - then, also saves an iteration in the recursion if there was a trailing slash when first submitted.